### ISSUE TITLE
Fails to install due to linking error "undefined reference to `SSL_ctrl"

### ISSUE THREAD

**Issue by Christian-Prather** [cargo-generate/cargo-generate#1266]
**Describe the bug**
Running `cargo install cargo-generate` fails to install due to the following error
```
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/christian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/christian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/christian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/christian/.pyenv/shims:/home/christian/.pyenv/bin:/home/christian/.local/bin:/home/christian/.nvm/versions/node/v16.15.1/bin:/home/christian/.cargo/bin:/home/christian/.rvm/gems/ruby-2.7.2/bin:/home/christian/.rvm/gems/ruby-2.7.2@global/bin:/home/christian/.rvm/rubies/ruby-2.7.2/bin:/home/christian/.local/bin:/home/christian/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/christian/.rvm/bin:/home/christian/mycroft-core/bin:/home/christian/.rvm/bin:/home/christian/Luxonis/depthai/entrypoint" VSLANG="1033" "cc" "-m64" "/tmp/rustcwNQ5sr/symbols.o" "/tmp/cargo-installqpurGb/release/deps/cargo_generate-b60766f0931fe120.cargo_generate.ab897d2350b21ae7-cgu.1.rcgu.o" "-Wl,--as-needed" "-L" "/tmp/cargo-installqpurGb/release/deps" "-L" "/tmp/cargo-installqpurGb/release/build/libgit2-sys-53c858e82307dda1/out/build" "-L" "/tmp/cargo-installqpurGb/release/build/libssh2-sys-1734404110029251/out/build" "-L" "/usr/local/lib" "-L" "/home/christian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/tmp/rustcwNQ5sr/liblibgit2_sys-fd1249554d39f385.rlib" "/tmp/rustcwNQ5sr/liblibssh2_sys-94dba8ef55a11810.rlib" "/home/christian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-6648dc218e522d87.rlib" "-Wl,-Bdynamic" "-lssl" "-lcrypto" "-lz" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/christian/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/tmp/cargo-installqpurGb/release/deps/cargo_generate-b60766f0931fe120" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-Wl,--strip-debug" "-nodefaultlibs"
  = note: /usr/bin/ld: /tmp/rustcwNQ5sr/liblibgit2_sys-fd1249554d39f385.rlib(cccc76fb850512eb-openssl.o): in function `openssl_certificate':
          openssl.c:(.text.openssl_certificate+0x2b): undefined reference to `SSL_get1_peer_certificate'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibgit2_sys-fd1249554d39f385.rlib(cccc76fb850512eb-openssl.o): in function `openssl_connect':
          openssl.c:(.text.openssl_connect+0x84): undefined reference to `SSL_ctrl'
          /usr/bin/ld: openssl.c:(.text.openssl_connect+0xfd): undefined reference to `SSL_get1_peer_certificate'
          /usr/bin/ld: openssl.c:(.text.openssl_connect+0x12e): undefined reference to `OPENSSL_sk_num'
          /usr/bin/ld: openssl.c:(.text.openssl_connect+0x1a6): undefined reference to `OPENSSL_sk_value'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibgit2_sys-fd1249554d39f385.rlib(cccc76fb850512eb-openssl.o): in function `git_openssl_stream_global_init':
          openssl.c:(.text.git_openssl_stream_global_init+0x56): undefined reference to `SSL_CTX_ctrl'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibssh2_sys-94dba8ef55a11810.rlib(903bc623e643396f-mac.o): in function `mac_method_hmac_ripemd160_hash':
          mac.c:(.text.mac_method_hmac_ripemd160_hash+0x4d): undefined reference to `EVP_ripemd160'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibssh2_sys-94dba8ef55a11810.rlib(903bc623e643396f-openssl.o): in function `_libssh2_ed25519_new_private_frommemory':
          openssl.c:(.text._libssh2_ed25519_new_private_frommemory+0x63): undefined reference to `EVP_PKEY_get_id'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibssh2_sys-94dba8ef55a11810.rlib(903bc623e643396f-openssl.o): in function `_libssh2_pub_priv_keyfile':
          openssl.c:(.text._libssh2_pub_priv_keyfile+0x9d): undefined reference to `EVP_PKEY_get_id'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibssh2_sys-94dba8ef55a11810.rlib(903bc623e643396f-openssl.o): in function `_libssh2_pub_priv_keyfilememory':
          openssl.c:(.text._libssh2_pub_priv_keyfilememory+0x86): undefined reference to `EVP_PKEY_get_id'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibssh2_sys-94dba8ef55a11810.rlib(903bc623e643396f-crypt.o):(.data.rel.ro.libssh2_crypt_method_cast128_cbc+0x40): undefined reference to `EVP_cast5_cbc'
          /usr/bin/ld: /tmp/rustcwNQ5sr/liblibssh2_sys-94dba8ef55a11810.rlib(903bc623e643396f-crypt.o):(.data.rel.ro.libssh2_crypt_method_blowfish_cbc+0x40): undefined reference to `EVP_bf_cbc'
          collect2: error: ld returned 1 exit status
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-link-lib)

error: could not compile `cargo-generate` (bin "cargo-generate") due to 1 previous error
error: failed to compile `cargo-generate v0.22.0`, intermediate artifacts can be found at `/tmp/cargo-installqpurGb`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.

```

**make sure you**
- [ x] I'm on the latest rust version `rustup update`
- [ x] I'm on the latest cargo-generate version `cargo install cargo-generate`
- [ x] In case you have an own `.gitconfig`, provide all sections that contains one of: `insteadOf`, `url`

**To Reproduce**
1. cargo install cargo-generate

**Expected behavior**
cargo generate to build and install correctly

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Linux 
 - Version: 22.04.4
 - Build tools: gnu
 - Shell: bash
 - `rustc --version`: rustc 1.80.1 (3f5fd8dd4 2024-08-06)
 - `cargo generate --version`: x
 - `cargo --version:`: cargo 1.80.1 (376290515 2024-07-16)

**Additional context**

**Comment by sassman** [comment]
do you have the `libssl-dev` installed?

**Comment by Christian-Prather** [comment]
Hi, yes I do sorry I should have included that in my initial report.

**Comment by sassman** [comment]
I took a fresh ubuntu 22.04 docker image and ran the following:

```
apt-get update
apt-get install curl gcc pkg-config libssl-dev
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
```

Then I was able to install cargo-generate

```
cargo install cargo-generate
```

**Comment by sassman** [comment]
> Hi, yes I do sorry I should have included that in my initial report.

No worries.

After a bit of googling, installing an updated version of the `libssl-dev` package might be a solution.

**Comment by sassman** [comment]
Also as alternative you could install `cargo-generate` via `cargo-binstall`

**Comment by Christian-Prather** [comment]
Odd, I am also able to install on the docker image? However `openssl version` yields the same output as on my dev system? Is there something with my rust installation? I have uninstalled and reinstalled with rustup but not change?

**Comment by Christian-Prather** [comment]
IVe tried reinstalling rust, clearing my .cargo registry file and installing other packages with cargo install. This is the only package that seems to fail?

**Comment by Christian-Prather** [comment]
If I run `cargo install cargo-generate --features vendored-openssl` it does install however?

**Comment by sassman** [comment]
> If I run `cargo install cargo-generate --features vendored-openssl` it does install however?

It makes sense, because the vendored version, is that cargo generate brings it's own openssl dependency. It does not rely on the system-installed version.

So I'm suggesting you uninstall `libssl-dev` and re-install it.

**Comment by Christian-Prather** [comment]
I uninstalled / reinstalled `libssl-dev` and apt reinstalled `openssl` neither made a change. I am still getting the same error as before. This seems to have corresponded with my adding a riscv target with rustup however I have since removed that target and there is no change.

**Comment by sassman** [comment]
But for what architecture you're building cargo-generate for? X86_64 or arm64?

**Comment by Christian-Prather** [comment]
X86 unless there is some config I am unaware of setting something besides the default?

**Comment by Christian-Prather** [comment]
Okay so resolved this by going nuclear and taking the chance to upgrade to Ubuntu 24.04 which I had been wanting to do for some time. I believe this was due to some issue in my rust / system libraries? I am wondering if it was due to my killing a build during the linking stage with ctrl-c as I found some other posts with that causing similar behavior to what I experienced. Thank you for all the help and sorry it ended up being a larger scope than this crate!
